src/unix/socket.c

```diff
@@ -0,0 +1,579 @@
+            child->connecting = false;
+            blockq_flush(child->sock.txbq);
+        }
+
+        deallocate_queue(s->conn_q);
+    }
+    if (s->fs_entry) {
+        buffer_clear(table_find(s->fs_entry, sym(socket)));
+    }
+    unixsock_dealloc(s);
+    return 0;
+}
+
+static sysreturn unixsock_bind(struct sock *sock, struct sockaddr *addr,
+        socklen_t addrlen)
+{
+    if (!addr) {
+        return -EFAULT;
+    }
+
+    unixsock s = (unixsock) sock;
+    struct sockaddr_un *unixaddr = (struct sockaddr_un *) addr;
+    if (s->fs_entry || (addrlen <= sizeof(unixaddr->sun_family))) {
+        return -EINVAL;
+    }
+    if (addrlen > sizeof(*unixaddr)) {
+        return -ENAMETOOLONG;
+    }
+
+    /* Ensure that the NULL-terminated path string fits in unixaddr->sun_path
+     * (add terminator character if not found). */
+    int term;
+    for (term = 1; term < addrlen - sizeof(unixaddr->sun_family); term++) {
+        if (unixaddr->sun_path[term] == '\0') {
+            break;
+        }
+    }
+    if (term == addrlen - sizeof(unixaddr->sun_family)) {
+        /* Terminator character not found: add it if possible. */
+        if (addrlen == sizeof(*unixaddr)) {
+            return -ENAMETOOLONG;
+        }
+        unixaddr->sun_path[term] = '\0';
+    }
+
+    s->fs_entry = allocate_tuple();
+
+    /* The "socket" symbol value associated to the filesystem tuple is a buffer
+     * containing a pointer to the socket. Since we don't want this pointer to
+     * be persisted in the filesystem, the buffer must be empty when
+     * filesystem_add_tuple() is called, and must be filled afterwards. */
+    buffer b = allocate_buffer(sock->h, sizeof(u64));
+    table_set(s->fs_entry, sym(socket), b);
+    sysreturn ret = filesystem_add_tuple(unixaddr->sun_path, s->fs_entry);
+    if (ret) {
+        deallocate_buffer(b);
+        deallocate_tuple(s->fs_entry);
+        s->fs_entry = 0;
+        if (ret == -EEXIST) {
+            return -EADDRINUSE;
+        }
+        else {
+            return ret;
+        }
+    }
+    buffer_write_le64(b, u64_from_pointer(s));
+
+    runtime_memcpy(&s->local_addr, addr, addrlen);
+    return ret;
+}
+
+static sysreturn unixsock_listen(struct sock *sock, int backlog)
+{
+    unixsock s = (unixsock) sock;
+    if (!s->conn_q) {
+        s->conn_q = allocate_queue(sock->h, backlog);
+        if (s->conn_q == INVALID_ADDRESS) {
+            msg_err("failed to allocate connection queue\n");
+            s->conn_q = 0;
+            return -ENOMEM;
+        }
+    }
+    return 0;
+}
+
+closure_function(2, 1, sysreturn, connect_bh,
+                 unixsock, s, thread, t,
+                 u64, bqflags)
+{
+    unixsock s = bound(s);
+    thread t = bound(t);
+    sysreturn rv;
+
+    if (bqflags & BLOCKQ_ACTION_NULLIFY) {
+        rv = -EINTR;
+        goto out;
+    }
+    if (!s->connecting) {   /* the listening socket has been shut down */
```

Comment:
Right, a NULLIFY flag doesn't automatically mean a signal has been raised, so -EINTR should be returned only if NULLIFY occurs because of a signal, and not if it occurs because of the peer socket being closed. To fix this, I added `&& s->connecting` to the `if` clause that checks for NULLIFY.
The same type of issue is also present in read_bh and write_bh, which I took from the previous socketpair implementation. I fixed those as well by adding `&& s->peer` to the `if` clauses that check for NULLIFY, so that the correct return value (0 for read and -EPIPE for write) is used instead of -EINTR when the peer socket has been closed.